src/plugin/yjsEditor.ts

```diff
@@ -60,40 +51,51 @@ export const YjsEditor = {
     fn();
 
     if (!wasRemote) {
-      Promise.resolve().then(() => IS_REMOTE.delete(editor));
+      IS_REMOTE.delete(editor);
     }
   },
+};
 
-  /**
-   * Apply Yjs events to slate
-   */
-  applyYjsEvents: (editor: YjsEditor, events: Y.YEvent[]): void => {
-    YjsEditor.asRemote(editor, () => {
-      applyYjsEvents(editor, events);
-    });
-  },
+function localOperations(editor: YjsEditor): Set<Operation> {
+  const operations = LOCAL_OPERATIONS.get(editor);
+  invariant(operations, 'YjsEditor without attached local operations');
+  return operations;
+}
 
-  /**
-   * Performs an action as a local operation.
-   */
-  asLocal: (editor: YjsEditor, fn: () => void): void => {
-    const wasLocal = YjsEditor.isLocal(editor);
-    IS_LOCAL.add(editor);
+function trackLocalOperations(editor: YjsEditor, operation: Operation): void {
+  if (!YjsEditor.isRemote(editor)) {
+    localOperations(editor).add(operation);
+  }
+}
 
-    fn();
+/**
+ * Applies a slate operations to the bound shared type.
+ */
+function applyLocalOperations(editor: YjsEditor): void {
+  const editorLocalOperations = localOperations(editor);
 
-    if (!wasLocal) {
-      IS_LOCAL.delete(editor);
-    }
-  },
+  applySlateOps(
+    YjsEditor.sharedType(editor),
+    Array.from(editorLocalOperations),
+    slateYjsSymbol
+  );
 
-  /**
-   * Returns whether the editor currently is applying a remote change to the yjs doc.
-   */
-  isLocal: (editor: YjsEditor): boolean => {
-    return IS_LOCAL.has(editor);
-  },
-};
+  editorLocalOperations.clear();
+}
+
+/**
+ * Apply Yjs events to slate
+ */
+function applyRemoteYjsEvents(editor: YjsEditor, events: Y.YEvent[]): void {
+  Editor.withoutNormalizing(editor, () =>
+    YjsEditor.asRemote(editor, () =>
```

Comment:
This is bugging me a bit, as it basically means anyone doing asRemote will need to remember to also do withoutNormalizing. What do you think about moving withoutNormalizing into asRemote (see diff)? It passes the tests so it should be good to go.

```diff
diff --git a/src/plugin/yjsEditor.ts b/src/plugin/yjsEditor.ts
index 25a7570..bf300f0 100644
--- a/src/plugin/yjsEditor.ts
+++ b/src/plugin/yjsEditor.ts
@@ -45,14 +45,16 @@ export const YjsEditor = {
    * Performs an action as a remote operation.
    */
   asRemote: (editor: YjsEditor, fn: () => void): void => {
-    const wasRemote = YjsEditor.isRemote(editor);
-    IS_REMOTE.add(editor);
+    Editor.withoutNormalizing(editor, () => {
+      const wasRemote = YjsEditor.isRemote(editor);
+      IS_REMOTE.add(editor);
 
-    fn();
+      fn();
 
-    if (!wasRemote) {
-      IS_REMOTE.delete(editor);
-    }
+      if (!wasRemote) {
+        IS_REMOTE.delete(editor);
+      }
+    });
   },
 };
 
@@ -87,12 +89,10 @@ function applyLocalOperations(editor: YjsEditor): void {
  * Apply Yjs events to slate
  */
 function applyRemoteYjsEvents(editor: YjsEditor, events: Y.YEvent[]): void {
-  Editor.withoutNormalizing(editor, () =>
-    YjsEditor.asRemote(editor, () =>
-      applyYjsEvents(
-        editor,
-        events.filter((event) => event.transaction.origin !== slateYjsSymbol)
-      )
+  YjsEditor.asRemote(editor, () =>
+    applyYjsEvents(
+      editor,
+      events.filter((event) => event.transaction.origin !== slateYjsSymbol)
     )
   );
 }
```

Reply:
Not a huge fan of this tbh since it means that asRemote influences the slate behavior in a way you might not expect. Maybe we should just put a warning in the jsdoc or something.

Reply:
Well, whichever you prefer. As for me i don't have anything else to add at this moment, so looking forward to the release :)